### PR TITLE
Additional parsing for scripts

### PR DESCRIPTION
**Pull request by Regent**
Whilst investigating original issue I've found a few more cases when parser would mistake special characters for end tag sequence.

### CODE REVIEW

File: src/Html/HtmlParser.fs [hunk 1]
```diff
@@ -423,8 +423,40 @@ module internal HtmlParser =
...
+            match state.Peek() with
+            | ''' -> state.Cons(); script state
+            | _ -> state.Cons(); scriptSingleQuoteString state
+        and scriptDoubleQuoteString state =
+            match state.Peek() with
+            | '"' -> state.Cons(); script state
+            | _ -> state.Cons(); scriptDoubleQuoteString state
+        and scriptSlash state =
+            match state.Peek() with
+            | '/' -> state.Cons(); scriptSinleLineComment state
```
Comment (ovatsus) [inline src/Html/HtmlParser.fs]: Typo: sinle instead of single

File: tests/FSharp.Data.Tests/HtmlParser.fs [hunk 2]
```diff
@@ -96,6 +96,18 @@ let ``Can handle attributes with no value``() =
         ]
     node.Attributes() |> should equal expected
 
+[<TestCase("var r =\"</script>\"")>]
+[<TestCase("var r ='</script>'")>]
+[<TestCase("var r =/</g")>]
```
Comment (ovatsus) [inline tests/FSharp.Data.Tests/HtmlParser.fs]: Can you add a test case with "var r =/\/</g"

File: tests/FSharp.Data.Tests/HtmlParser.fs [hunk 3]
```diff
@@ -96,6 +96,18 @@ let ``Can handle attributes with no value``() =
...
+[<TestCase("//</script>\n")>]
```
Comment (ovatsus) [inline tests/FSharp.Data.Tests/HtmlParser.fs]: Can you add the opposite case, not having the \n at the end to make sure it fails

File: tests/FSharp.Data.Tests/HtmlParser.fs [hunk 4]
```diff
@@ -96,6 +96,18 @@ let ``Can handle attributes with no value``() =
...
+[<TestCase("/*</script>*/")>]
+[<TestCase("/*</script>**/")>]
```
Comment (ovatsus) [inline tests/FSharp.Data.Tests/HtmlParser.fs]: Can you add test cases that use `/* */` and span multiple lines?

**Comment by ovatsus** [comment]
@Regent , is this finished now or are you still looking at more corner cases? @colinbull , can you have a quick look as well?

**Comment by Regent** [comment]
@ovatsus, in coming days I also want to cover the case of parsing `<script>0;<body></body>` correctly - at the moment it would parse as simply `<script />` - but ideally I would like to send a separate PR for that.

**Comment by ovatsus** [comment]
Sorry, I thought I had already merged this before

**Comment by ovatsus** [comment]
Released in 2.4.3

**Comment by glchapman** [comment]
Doing some scraping, I found a case where it appears this change (or at least changes since 2.4.2) do not find the end tag of a script. The url in question is :
https://www.marketwatch.com/investing/stock/msft/analystestimates
(where any stock ticker can replace "msft").

I'm attaching captures from the VS Debugger's text visualization of the script node's contents (Raw View).  As you can see, in 2.4.2, the script node is correctly terminated, whereas in 2.4.6, it is not terminated until the closing tag of the ensuing script.

[Script_2.4.2.txt](https://github.com/fsharp/FSharp.Data/files/1936017/Script_2.4.2.txt)
[Script_2.4.6.txt](https://github.com/fsharp/FSharp.Data/files/1936026/Script_2.4.6.txt)

**Comment by glchapman** [comment]
I believe I've narrowed down my problem to the handling of regexes in scripts. As you can see in the attached example, the regex contains a '/' in a bracketed character class. The tokenizer mistakes this for the end delimiter of the regex, and so fails to tokenize correctly.
[test_html.txt](https://github.com/fsharp/FSharp.Data/files/1939269/test_html.txt)
Here's fsi output for 2.4.6 with the above test.html:
```
> let doc = HtmlDocument.Load("test.html");;
val doc : HtmlDocument =
  <html>
  <head />
  <body>
    <script type="text/javascript">
  function(selector){
    return selector.replace(/([/.])/g, '\\$1');
  };
</script>
<div>captured by script</div>
<script></script>
  </body>
</html>

> let script = HtmlDocument.descendantsNamed false ["script"] doc |> Seq.head;;
val script : HtmlNode =
  <script type="text/javascript">
  function(selector){
    return selector.replace(/([/.])/g, '\\$1');
  };
</script>
<div>captured by script</div>
<script></script>
```

**Comment by ovatsus** [comment]
@glchapman , can you send a PR with a failing test case please? And if you could also fix it that would be great, but if you send the failing tests that's already very helpful

**Comment by Regent** [comment]
I haven't realised that you don't need to escape forward slash when it is inside a group. I can't find a definitive specification on the matter - hoped to get it from www.ecma-international.org but they might have forgotten to pay for the hosting...

**Comment by glchapman** [comment]
I'll try to send a PR with a failing case in a bit, but FWIW my reading of the current HTML standard implies you don't actually have to parse the script text (to handle regexes, etc). Instead, raw text nodes (the content of scripts and styles) can have any text at all up to the terminating tag:
https://www.w3.org/TR/2017/REC-html52-20171214/syntax.html#restrictions-on-the-contents-of-raw-text-and-escapable-raw-text-elements
See also the description of parsing the RAWTEXT state:
https://www.w3.org/TR/2017/REC-html52-20171214/syntax.html#rawtext-state

I don't know if this differs from earlier HTML variants.

**Comment by colinbull** [comment]
I'm just looking at this now. @glchapman Can you let me push to your PR with the failing test?

**Comment by glchapman** [comment]
@colinbull  -- I've added you as a collaborator to my fork with the PR. Is this what you need? (I'm sorry, I'm not a big github user).

**Comment by colinbull** [comment]
Yep thanks. Ignore my comments before I have caught up on whats going on now.

**Comment by Regent** [comment]
Original motivation was to handle some corner-cases, e.g. so that </ within regex doesn't get parsed as start of an end tag.

But that's a good point - do we need to parse scripts at all. If I read it correctly, HTML spec simply looks for a closing script tag.
One good thing about current implementation is that it's resilient to the lack of closing tag.

**Comment by colinbull** [comment]
Ok, I have the unit tests passing now but have broken the signature tests (ebay_cars.htm). I haven't got anymore time tonight to look but hoping to get back to this, tomorrow or wednesday.
